### PR TITLE
Add file operation coordinator to the registry

### LINKED ISSUE #13183: Add property-based editor testing with fast-check
## Goal

Introduce `fast-check` as the foundation for property-based testing around the editor-buffer refactor.

This should land early so later buffer, descriptor, persistence, execution, snapshot, and project-action work starts with generative coverage instead of adding it afterward.

## Acceptance Criteria

- `fast-check` is available in the existing TypeScript test environment.
- Test helpers report enough seed/path information to replay failures deterministically.
- Initial properties cover path normalization, descriptor inputs, deterministic capability ordering/deduplication, save coalescing, and stale async result rejection.
- Reusable state-machine test models can generate open/edit/save/rename/delete and view mount/unmount/remount sequences against persistent buffers.
- Helpers can generate multi-buffer version maps and prepared workspace mutations for later atomicity/conflict tests.
- Property tests stay focused on pure reducers/helpers and do not require Electron, a browser, or a live engine connection.
- The testing pattern is documented well enough for #13184, #13185, #13353, and #13354 to reuse.

## Notes

This task is intentionally small and may land before the project-session/fs-operation-queue stack is fully merged.

### PR DESCRIPTION
**Pull request by franknoirot**
Stacked on #13597, towards #12350, closes #13183. This adds `FileOperations` as what will become the sole application-facing filesystem authority. It uses [Effect pipelines](https://effect.website/docs/v4/getting-started/building-pipelines) to coordinate file system operations robustly.

## What the `FileOperations` Effect pipeline does

1. Adds coordinated access checks, stat, existence, directory and file reads, writes, copies, moves, renames, removals, and strict or unique file/directory creation.
2. Accepts either `string` or `Uint8Array` contents for every file-writing operation, encoding strings as UTF-8 and snapshotting mutable bytes before they enter the queue.
3. Uses shared/exclusive path locks so related paths serialize without blocking unrelated paths.
4. Adds directory-membership locks so listings cannot observe a child being added, removed, or renamed halfway through an operation.
5. Pipes through the structured errors from `FileSystem` underneath, and adds additional client error reporting around directory and file lock failures (see "Client error reporting" below).

The `FileOperations` system is also registered as a core extension called `fileOperations` in this PR. As we complete our "strangler fig" migration it will become the only place callers are allowed to access the file system, enforced by lint rules (the next PR #13630 introduces this).

## Coordination guarantees

One of the core problems using Effect helps solve for us is *path locking*. This allows us to parallelize IO operations that are unrelated while strictly ordering ones that depend on each other: it's exactly what Effect pipelines are for. Here are some of the guarantees from the locks in this PR:

1. Access checks, stat, existence checks, and file reads coordinate with mutations of the same path.
2. Directory reads hold a shared membership lock while taking a names-only snapshot; child creation, removal, and rename hold the corresponding exclusive membership lock. Consumers obtain entry metadata through coordinated `stat` calls according to their own error policy.
3. A parent-directory rename, move, or removal waits for coordinated reads and writes beneath that directory, and descendant operations wait while the parent mutation is in progress.
4. `move` first attempts rename and falls back to copy-and-remove for cross-device moves while retaining the source/destination locks for the whole compound operation.
5. Acknowledged writes to the same path are serialized, so a later completion cannot be overwritten by an older queued write.
6. Writes to existing sibling files can still proceed concurrently. A write that creates a new path briefly participates in the parent membership lock.
7. Strict creation fails on an existing path; unique creation selects a non-colliding numbered name within this runtime's coordination boundary.
8. Multi-path operations acquire stable lock plans to avoid deadlocks.

These guarantees apply to callers using this mounted `FileOperations` runtime. They do not claim atomicity against another browser tab, process, or backend writer.

## Client error reporting

A rejected `FileOperations` request emits one `file_operations_error` report; internal `FileSystem` work does not emit a second report. The report records the coordinated operation, underlying filesystem operation, typed error, and allowlisted backing error code/name.

Paths, messages, stacks, filenames, and contents are excluded for user privacy. The reporter also supplies a fixed `file-operations` route instead of inheriting the current application route, because file routes may contain encoded absolute paths. `FileAlreadyExists` from strict `createFile` or `createDirectory` is intentionally not reported because callers use that result for ordinary collision feedback.

## `fast-check` is introduced in this PR for property testing / fuzzing

`fast-check` is specifically aimed at testing broad categories of failures by generating test inputs and expectations based on ["properties"](https://fast-check.dev/docs/tutorials/quick-start/our-first-property-based-test/#what-is-a-property). This is great for [detecting and preventing race conditions](https://fast-check.dev/docs/tutorials/detect-race-conditions/), which have plagued our IO systems. They have a `*.properties.*` file name.

The property tests in this PR generate Unicode strings, operation streams, and path sets rather than enumerating only hand-picked examples. Each generated case is checked against observable invariants, and a failure is [shrunk](https://github.com/dubzzz/fast-check/issues/97) to a smaller reproducible counterexample. Conventional tests remain alongside them as readable examples.

### CODE REVIEW

File: src/lib/fileSystem/fileNames.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/** Naming parts used to place a uniqueness suffix before the full extension. */
+export interface FileNameParts {
+  readonly stem: string
+  readonly extension: string
+}
+
+/** Build the preferred filename or its numbered collision variant. */
+export function fileNameCandidate(name: FileNameParts, suffix: number): string {
```
Comment (franknoirot) [inline src/lib/fileSystem/fileNames.ts:8]: I know this is duplicative of our path helpers, but idk I kinda want this to stand alone? Or I need our path libraries to have less baggage and use only pure functions

File: src/registry/contracts/fileOperations.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import type { FileNameParts } from '@src/lib/fileSystem/fileNames'
+import type {
+  CopyOptions,
+  DirectoryEntry,
+  FileContents,
+  FileStat,
+} from '@src/lib/fileSystem/fileOperations'
+
+/** Promise facade for coordinated project-directory and file operations. */
+export interface FileOperationsRegistryService {
```
Comment (franknoirot) [inline src/registry/contracts/fileOperations.ts:10]: This is where to start: this service should have everything we need to interact with a file operation queue in one shared place, so let me know if anything looks wrong.

File: src/lib/fileSystem/runtime.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    program: Effect.Effect<A, E, R>,
+    onFailure?: (error: E) => void
+  ): Promise<A> => {
+    const result = await runtime.runPromise(program.pipe(Effect.either))
+    if (Either.isLeft(result)) {
+      onFailure?.(result.left)
+      return Promise.reject(result.left)
+    }
+    return result.right
+  }
```
Comment (franknoirot) [inline src/lib/fileSystem/runtime.ts:64]: Effect calls nested operations "programs" and they can compose really nicely in reusable ways, since [it's just a bougie functional programming library](https://dev.to/effect/a-bright-future-for-effect-455m). `Effect.either` is a reusable program they export that picks left (unsuccessful) or right (successful) outcomes, like lambda calculus.

Each of the service methods is just wrapped in this so they all handle rejection and piping in the same way.

File: src/lib/fileSystem/runtime.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+  const runFileOperationsPromise = <A>(
+    operation: FileOperationsOperation,
+    program: Effect.Effect<A, FileSystemError, FileOperations>
+  ) =>
+    runRuntimePromise(program, (error) =>
+      reportFileOperationsError(operation, error)
```
Comment (franknoirot) [inline src/lib/fileSystem/runtime.ts:71]: This is where the unsuccessful case's return type is established, so Effect can tell the TypeScript type checker what error types are possible for each of these service methods!

File: src/lib/fileSystem/pathLocking.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    while (true) {
+      const parent = backing.dirname(current)
+      if (parent === current) {
+        break
+      }
+
+      if (!modesByPath.has(parent)) {
+        modesByPath.set(parent, 'shared')
+      }
```
Comment (franknoirot) [inline src/lib/fileSystem/pathLocking.ts:40]: This is where we detect descendants for lock status

File: src/lib/fileSystem/pathLocking.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ */
+export interface PathLockRequirement {
+  readonly path: string
+  readonly mode: 'shared' | 'exclusive'
+}
+
+/**
+ * Builds a stable, duplicate-free lock plan for one or more mutation targets.
+ */
+export function pathLockRequirements(
```
Comment (franknoirot) [inline src/lib/fileSystem/pathLocking.ts:18]: This is called "requirements" because the Effect type is `Effect<A, E, R>`, with `R` standing for Requirements: the context needed to run the effect.

File: src/lib/fileSystem/pathLocking.properties.test.ts [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+const segmentArbitrary = fc
+  .array(fc.constantFrom(...'abcdefghijklmnopqrstuvwxyz0123456789-_'), {
+    minLength: 1,
+    maxLength: 12,
+  })
+  .map((characters) => characters.join(''))
+
+const pathArbitrary = fc
+  .array(segmentArbitrary, { minLength: 1, maxLength: 5 })
+  .map((segments) => nodeFileSystem.impl.join('workspace', ...segments))
```
Comment (franknoirot) [inline src/lib/fileSystem/pathLocking.properties.test.ts:18]: Think of these are random generators with rules. They're used in the tests below to create random arbitrary file paths with up to 5 segments.

File: src/lib/fileSystem/fileOperations.ts [hunk 8]
```diff
@@ -0,0 +1,523 @@
...
+  readonly overwrite?: boolean
+}
+
+const textEncoder = new TextEncoder()
+
+function snapshotFileContents(contents: FileContents): Uint8Array {
+  return typeof contents === 'string'
+    ? textEncoder.encode(contents)
+    : new Uint8Array(contents)
+}
```
Comment (franknoirot) [inline src/lib/fileSystem/fileOperations.ts:51]: UInt8Arrays are mutable so we have to snapshot them to make sure they're not changed while we're locking somehow.

File: src/lib/fileSystem/fileOperations.ts [hunk 9]
```diff
@@ -0,0 +1,523 @@
...
+  readonly rename: (
+    source: string,
+    destination: string
+  ) => Effect.Effect<void, FileSystemError>
+}
+
+export class FileOperations extends Context.Tag('zds/FileOperations')<
+  FileOperations,
+  FileOperationsService
+>() {}
```
Comment (franknoirot) [inline src/lib/fileSystem/fileOperations.ts:139]: The "tag" is just a way to reference the layer in Effect (more for types and stuff), but the actual layer is called `fileOperationsLayer` below.

File: src/lib/fileSystem/fileOperations.ts [hunk 10]
```diff
@@ -0,0 +1,523 @@
...
+
+const makeFileOperations = (backing: IZooDesignStudioFS) =>
+  Effect.gen(function* () {
+    const fileSystem = yield* FileSystem
+    const pending = yield* SubscriptionRef.make(0)
+    const pathLocks = yield* SynchronizedRef.make(
+      new Map<string, PathLockEntry>()
+    )
```
Comment (franknoirot) [inline src/lib/fileSystem/fileOperations.ts:147]: Effect uses generators to let us write code that (mostly) looks like normal JS but with the powerful functional composition and error tagging from Effect underneath. Check out [this guide](https://effect.website/docs/v4/getting-started/using-generators#understanding-effectgen) to get familiar with it.

File: src/lib/fileSystem/fileOperations.ts [hunk 11]
```diff
@@ -0,0 +1,523 @@
...
+        reserveLock(locks, requirement.path),
+        (semaphore) =>
+          semaphore.withPermits(
+            requirement.mode === 'exclusive'
+              ? EXCLUSIVE_LOCK_PERMITS
+              : SHARED_LOCK_PERMITS
+          )(withLocks(locks, requirements, operation, index + 1)),
+        () => releaseLock(locks, requirement.path)
+      )
+    }
```
Comment (franknoirot) [inline src/lib/fileSystem/fileOperations.ts:207]: This right here is why I reached for Effect. Locking and coordination is *hard as fuck*, but this library lets us compose those policies that, at least at the end of the pipeline, are pretty readable, and way easier to debug. Plus it has built-in notions of parallelization and semaphores.

File: src/lib/fileSystem/fileOperations.ts [hunk 12]
```diff
@@ -0,0 +1,523 @@
...
+      paths: readonly string[],
+      operation: Effect.Effect<A, E, R>,
+      membershipPaths = paths.map((path) => backing.dirname(path))
+    ) =>
+      trackMutation(
+        withPathLocks(
+          pathLockRequirements(backing, paths),
+          withDirectoryMembershipLocks(membershipPaths, 'exclusive', operation)
+        )
+      )
```
Comment (franknoirot) [inline src/lib/fileSystem/fileOperations.ts:258]: See now we can see here that mutations are coordinated by applying exclusive membership locks to all the necessary paths.

File: src/lib/fileSystem/fileOperations.ts [hunk 13]
```diff
@@ -0,0 +1,523 @@
...
+          [parent]
+        ),
+      remove: (path) => coordinateMutation([path], fileSystem.remove(path)),
+      rename: (source, destination) =>
+        coordinateMutation(
+          [source, destination],
+          fileSystem.rename(source, destination)
+        ),
+    })
+  })
```
Comment (franknoirot) [inline src/lib/fileSystem/fileOperations.ts:419]: This is the part I hope is most readable.

File: src/lib/fileSystem/fileOperations.properties.test.ts [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+  )
+
+const decodeVersion = (contents: Uint8Array): VersionedContents =>
+  JSON.parse(new TextDecoder().decode(contents)) as VersionedContents
+
+/**
+ * Model the guarantees owned by one coordinator runtime. A backing adapter
+ * must separately guarantee that an individual successful write is atomic;
+ * cross-runtime ordering requires a shared platform or server authority.
+ */
```
Comment (franknoirot) [inline src/lib/fileSystem/fileOperations.properties.test.ts:58]: This means that we can only prove that this shit works on our current browser tab or desktop app, not that coordination occurs across clients. Although cool thing: since this uses an logical clock for versions, once a STUN server owns that *we could have live collab*.

File: src/lib/fileSystem/fileOperations.properties.test.ts [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+            ...new TextEncoder().encode(contents),
+          ])
+        }
+      ),
+      { numRuns: 50 }
+    )
+  })
+
+  it(
+    'preserves write order and never reads an older acknowledged revision',
```
Comment (franknoirot) [inline src/lib/fileSystem/fileOperations.properties.test.ts:102]: This is the big important data loss property test.

**Review by franknoirot — COMMENTED** [review]
Okay I know that adding both fast-check and Effect in this stack is a lot but I hope I can convince y'all it's worth it here.

File: src/lib/fileSystem/fileOperations.ts [hunk 16]
```diff
@@ -0,0 +1,521 @@
...
+            ? Effect.fail(
+                fileSystemError(
+                  operation,
+                  path,
+                  Object.assign(new Error(`Path ${path} already exists`), {
+                    code: 'EEXIST',
+                  })
+                )
+              )
+            : create
```
Comment (andrewvarga) [inline src/lib/fileSystem/fileOperations.ts:297]: Just 2 notes here:
- Personally I'm not a big fan of passing functions (effects technically here) as parameters because they make the code harder to reason about when reading it. The possible value of that parameter is dynamic and can be different from multiple call sites, so when I'm reading the code I have to look up the callers to see what it can resolve to in practice.. which is why its flexible so I understand its appeal. I would probably have 2 separate functions here for clarity (one for files and one for directories)

- The other is a small race possibility that we can probably live with but we could also fix: it's possible that after `fileSystem.exists(path)` returns that the path doesn't exist yet, an external process writes to that path exactly then, and then our own create call fails. This should be super hard to reproduce in practice, but we could ensure it never happens by adding a new createFile operation to FileSystem which fails if the path already exists, then createFileAt would call that instead of checking if the path is available.

File: src/lib/fileSystem/pathLocking.properties.test.ts [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+    if (
+      requirement.mode === 'exclusive' ||
+      !modesByPath.has(requirement.path)
+    ) {
+      modesByPath.set(requirement.path, requirement.mode)
+    }
+  }
+
+  return [...modesByPath]
+    .sort(([left], [right]) => left.localeCompare(right))
```
Comment (andrewvarga) [inline src/lib/fileSystem/pathLocking.properties.test.ts:54]: Another edge-case but this `localeCompare` should probably become string compare instead, something like:
```
.sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
```

`modesByPath` stores by string values and with some unicode characters it's possible that 2 different string values compare equal with `localCompare` even though they are different: `ab`, `a\u200bb`, (`\u200b` is an invisible character).

File: src/lib/fileSystem/fileOperations.ts [hunk 18]
```diff
@@ -0,0 +1,521 @@
...
+      // As with unique directories, an exclusive parent lock makes candidate
+      // selection and creation one coordinated operation.
+      createUniqueFile: (parent, name, contents) =>
+        coordinateMutation(
+          [parent],
+          createUniqueFileAt(parent, name, snapshotFileContents(contents)),
+          [parent]
+        ),
+      createDirectory: (path) =>
+        coordinateMutation([path], createDirectoryAt(path)),
```
Comment (andrewvarga) [inline src/lib/fileSystem/fileOperations.ts]: For my clarity:
- we now have path vs membership locks: path protects a path, membership protects the entries (children) of a directory.
- shared vs exclusive locks: reading is shared, exclusive for writing

This sequence is problematic:
- `/projects` exists and is empty.
- actor A starts `readDirectory('/projects')`, acquiring its shared membership lock - the filesystem listing is still in progress.
- actor B starts `createDirectory('/projects/chair/parts')`. It locks membership of - `/projects/chair`, but not `/projects`.
- B’s recursive mkdir creates `/projects/chair` while A is still listing /projects.
- B creates `parts` inside `chair` and finishes.

So the issue is that during directory listing we shouldn't modify the directory, but we do with the recursive mkdir, the fix would be to add an exclusive membership lock on `/projects`.

This would be hard to reproduce with anything user facing and may be not worth the effort for now, just mentioning for completeness.

File: src/lib/fileSystem/fileOperations.ts [hunk 19]
```diff
@@ -0,0 +1,521 @@
...
+      operations.copy(source, destination, options)
+    )
+  )
+
+export const move = (source: string, destination: string) =>
+  FileOperations.pipe(
+    Effect.flatMap((operations) => operations.move(source, destination))
+  )
+
+export const writeFile = (path: string, contents: FileContents) =>
```
Comment (andrewvarga) [inline src/lib/fileSystem/fileOperations.ts]: Similar to https://github.com/KittyCAD/modeling-app/pull/13597#discussion_r4003928272

File: src/lib/fileSystem/pathLocking.properties.test.ts (hunk 17 above)
Comment (franknoirot) [reply src/lib/fileSystem/pathLocking.properties.test.ts]: This was completed here, thanks for catching it `localeCompare` is way too loosey goosey https://github.com/KittyCAD/modeling-app/pull/13628/changes/ff2e88178636fcfe5ad34bdcb58986d1cde65b57#diff-d9f9e8fae71c4fb65c17e6e87384afb37ccab3c67c64afd3b02c5cd48127f2a3R15-R19

File: src/lib/fileSystem/fileOperations.ts (hunk 16 above)
Comment (franknoirot) [reply src/lib/fileSystem/fileOperations.ts]: The `fileSystem.exists(path)` issue was fixed upstream in #13597 here https://github.com/KittyCAD/modeling-app/pull/13597/changes/67f2e4de976cc236332bd6b3e8cb2eee49c58aed

Comment (franknoirot) [reply src/lib/fileSystem/fileOperations.ts]: And your first point was addressed in https://github.com/KittyCAD/modeling-app/commit/fa79dadd97 by making `createFile` and `createDirectoryAt`

File: src/lib/fileSystem/fileOperations.ts (hunk 18 above)
Comment (franknoirot) [reply src/lib/fileSystem/fileOperations.ts]: Codex solved this a bit differently here https://github.com/KittyCAD/modeling-app/pull/13628/changes/fa79dadd970bedbc7f34e236922f7e10306bef38#diff-8e79a0ffabc9152e03a72ed71c685e0b3085d4581bae5df90143069e9ec7941eL280-R319, trying to avoid forcing unrelated directory descendants to lock each other, so it introduced `directoryLineage` and `ensureDirectoryPath`. With those, your example flows like this:
1. Creating `/projects/chair` requires an exclusive membership lock on `/projects`.
2. That waits for `readDirectory('/projects')` to release its shared membership lock.
3. Creating `/projects/chair/parts` then takes the exclusive membership lock on `/projects/chair`.

Comment (andrewvarga) [reply src/lib/fileSystem/fileOperations.ts]: Cool! One small issue still, in another location is in move:
https://github.com/KittyCAD/modeling-app/blob/e24b37adbe3b114cf2b8a5cb254ec38801c919bf/src/lib/fileSystem/fileOperations.ts#L378

seems to bypass the new helper.. but I'm approving regardless and let you decide, these are pretty rare races.

**Review by andrewvarga — APPROVED** [review]

File: src/lib/fileSystem/pathLocking.properties.test.ts (hunk 7 above)
Comment (pierremtb) [reply src/lib/fileSystem/pathLocking.properties.test.ts]: Oh nice

File: package.json [hunk 20]
```diff
@@ -217,6 +218,7 @@
     "eslint-plugin-testing-library": "^7.16.2",
     "eslint-plugin-vitest": "^0.5.4",
     "fake-indexeddb": "^6.2.5",
+    "fast-check": "^4.9.0",
```
Comment (pierremtb) [inline package.json]: LFG
